editor/plugins/node_3d_editor_plugin.cpp

```diff
@@ -2941,16 +2944,235 @@ void Node3DEditorViewport::_notification(int p_what) {
 
 				geometry->clear_surfaces();
 				geometry->surface_begin(Mesh::PRIMITIVE_LINES);
-				geometry->surface_add_vertex(start_pos);
-				geometry->surface_add_vertex(end_pos);
+
+				Vector3 center = (start_pos + end_pos) / 2;
+				Vector3 camera_dir = (camera->get_transform().origin - center).normalized();
+				real_t offset_distance = 0.01;
+
+				geometry->surface_add_vertex(start_pos + camera_dir * offset_distance);
+				geometry->surface_add_vertex(end_pos + camera_dir * offset_distance);
 				geometry->surface_end();
 
+				geometry_xray->clear_surfaces();
+				geometry_xray->surface_begin(Mesh::PRIMITIVE_LINES);
+				geometry_xray->surface_add_vertex(start_pos);
+				geometry_xray->surface_add_vertex(end_pos);
+				geometry_xray->surface_end();
+
 				float distance = start_pos.distance_to(end_pos);
+				if (distance < 0.001) {
+					distance = 0.0;
+				}
 				ruler_label->set_text(TS->format_number(vformat("%.3f m", distance)));
 
-				Vector3 center = (start_pos + end_pos) / 2;
 				Vector2 screen_position = camera->unproject_position(center) - (ruler_label->get_custom_minimum_size() / 2);
 				ruler_label->set_position(screen_position);
+
+				bool show_components = Input::get_singleton()->is_key_pressed(Key::SHIFT);
+
+				if (show_components) {
+					ImmediateMesh *triangle_mesh = Object::cast_to<ImmediateMesh>(ruler_triangle_lines->get_mesh().ptr());
+					ImmediateMesh *triangle_mesh_xray = Object::cast_to<ImmediateMesh>(ruler_triangle_lines_xray->get_mesh().ptr());
+
+					if (triangle_mesh && triangle_mesh_xray) {
+						Vector3 delta = end_pos - start_pos;
+
+						const real_t threshold = 0.001;
+						if (Math::abs(delta.x) < threshold) {
+							delta.x = 0.0;
+						}
+						if (Math::abs(delta.y) < threshold) {
+							delta.y = 0.0;
+						}
+						if (Math::abs(delta.z) < threshold) {
+							delta.z = 0.0;
+						}
+
+						Vector3 corner_point;
+
+						if (Math::abs(delta.x) > 0.0 && Math::abs(delta.y) > 0.0 && Math::abs(delta.z) == 0.0) {
+							// XY plane
+							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
+						} else if (Math::abs(delta.x) > 0.0 && Math::abs(delta.z) > 0.0 && Math::abs(delta.y) == 0.0) {
+							// XZ plane
+							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
+						} else if (Math::abs(delta.y) > 0.0 && Math::abs(delta.z) > 0.0 && Math::abs(delta.x) == 0.0) {
+							// YZ plane
+							corner_point = Vector3(start_pos.x, start_pos.y, end_pos.z);
+						} else if (Math::abs(delta.x) > 0.0 && Math::abs(delta.y) > 0.0 && Math::abs(delta.z) > 0.0) {
+							// All three axes
+							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
+						} else {
+							corner_point = end_pos;
+						}
```

Comment:
```suggestion
						if (Math::abs(delta.x) > 0.0 && Math::abs(delta.y) > 0.0 && delta.z == 0.0) {
							// XY plane
							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
						} else if (Math::abs(delta.x) > 0.0 && Math::abs(delta.z) > 0.0 && delta.y == 0.0) {
							// XZ plane
							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
						} else if (Math::abs(delta.y) > 0.0 && Math::abs(delta.z) > 0.0 && delta.x == 0.0) {
							// YZ plane
							corner_point = Vector3(start_pos.x, start_pos.y, end_pos.z);
						} else if (Math::abs(delta.x) > 0.0 && Math::abs(delta.y) > 0.0 && Math::abs(delta.z) > 0.0) {
							// All three axes
							corner_point = Vector3(end_pos.x, start_pos.y, start_pos.z);
						} else {
							corner_point = end_pos;
						}
```
If any of them equaled -0.0, then the threshold checks would've flipped the sign.